src/functionalTest/java/uk/gov/hmcts/reform/bulkscanprocessor/controllers/WhenProcessingEnvelopeTest.java

```diff
@@ -0,0 +1,84 @@
+package uk.gov.hmcts.reform.bulkscanprocessor.controllers;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.microsoft.azure.storage.StorageException;
+import com.microsoft.azure.storage.blob.CloudBlobClient;
+import com.microsoft.azure.storage.blob.CloudBlobContainer;
+import com.microsoft.azure.storage.blob.CloudBlockBlob;
+import com.microsoft.azure.storage.core.PathUtility;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.web.client.RestTemplate;
+import uk.gov.hmcts.reform.bulkscanprocessor.config.StorageConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(SpringRunner.class)
+@ContextConfiguration(classes = {StorageConfiguration.class})
+public class WhenProcessingEnvelopeTest {
+
+    @Value("${test-url}")
+    private String testUrl;
+
+    @Value("${scan.delay}")
+    private int scanDelay;
+
+    private String serviceName = "test";
+
+    @Autowired
+    private CloudBlobClient cloudBlobClient;
+
+    private CloudBlobContainer testContainer;
+
+    @Before
+    public void setUp() throws URISyntaxException, StorageException {
+        String sasToken = obtainSASToken();
+        URI containerUri = cloudBlobClient.getContainerReference(serviceName).getUri();
+        testContainer = new CloudBlobContainer(PathUtility.addToQuery(containerUri, sasToken));
+    }
+
+    @Test
+    public void should_delete_zip_file_after_successful_ingestion()
+        throws URISyntaxException, StorageException, IOException {
+
+        // upload the file
+        String zipName = "4_24-06-2018-00-00-00.zip";
+        String zipPath = new File("src/integrationTest/resources/" + zipName).getAbsolutePath();
+
+        CloudBlockBlob blockBlobReference = testContainer.getBlockBlobReference(zipName);
+        blockBlobReference.uploadFromFile(zipPath);
+
+        waitForBlobProcess();
+
+        assertThat(blockBlobReference.exists()).isFalse();
+    }
+
+    private String obtainSASToken() {
+        RestTemplate restTemplate = new RestTemplate();
+        ObjectNode objectNode = restTemplate.getForObject(testUrl + "/token/" + serviceName, ObjectNode.class);
+        JsonNode sasToken = objectNode.get("sas_token");
+        if (sasToken != null) {
+            return sasToken.asText();
+        }
+        throw new RuntimeException("Empty SAS token");
+    }
+
+    private void waitForBlobProcess() {
+        try {
+            Thread.sleep(scanDelay + 1000); // extra second for a good measure...
```

Comment:
If there are no blobs then previous jobs should not take much time time to execute . We can just wait for say 5 seconds then check blob storage ?

Reply:
Or, if there's a way of polling with an eventual timeout, that would be ideal.

Reply:
By default `scan.delay` is set to 30 seconds. This is a period that begins after the task was finished.

1. file is uploaded
2. here is the rest of the delay after last executio
3. blob processor is run
4. after successful processing, the file is deleted
5. fixed scan delay is reset

It is possible that the file gets uploaded during task execution or a second after it, and the file will be processed with the next run. Hence waiting for a fixed duration may give false-negative results, i.e. the file was not deleted because the test wasn't "patient" enough and didn't give application a chance to run the process.

I think it would be safe to assume that when file is deleted after a period shorter than `scan.delay`, the file was successfully processed and deleted as part of processing flow.

Reply:
What I meant was that in order to save those tests time, instead of waiting for the maximum possible amount of time, do it in the "wait at most N seconds for this condition to be met, polling every N milliseconds" style, which is possible here.

Reply:
to clarify, here's a sample code that Leszek is talking about:
```
await()
  .atMost(30, SECONDS)
   .pollDelay(500, MILLISECONDS)
   .until(() -> checkWhetherSomethingExpectedHappened());
```
https://github.com/hmcts/private-beta-invitation-service/blob/master/src/functionalTest/java/uk/gov/hmcts/reform/pbis/e2e/SmokeTest.java#L57